storage/innobase/lock/lock0lock.cc

```diff
@@ -495,16 +495,63 @@ void lock_sys_t::close()
 }
 
 #ifdef WITH_WSREP
+
+/** Check if a high priority (BF) trx has to wait for the current
+lock holder based on Wsrep transaction state relations.
+
+This code resembles the one in `wsrep_handle_mdl_conflict()`, but
+it's specific to the storage engine and it doesn't perform any
+BF aborts by itself.
+The decision whether to BF abort a victim may depend on other conditions
+like lock compatibility between InnoDB transactions.
+
+@param lock  other waiting lock
+@param trx   BF transaction requesting conflicting lock
+@return 0 if trx doesn't have to wait
+        1 if trx has to wait
+        -1 if a (potentially invalid) BF-BF wait case
+*/
+static int wsrep_BF_has_to_wait(const lock_t *lock, const trx_t *trx)
+{
+  THD *request_thd= trx->mysql_thd;
+  THD *granted_thd= lock->trx->mysql_thd;
+
+  /* Assume requester is BF. */
+  DBUG_ASSERT(wsrep_thd_is_BF(request_thd, false));
+
+  /* Granted is aborting, let it complete. */
+  if (wsrep_thd_is_aborting(granted_thd))
+      return 1;
+
+  /* Wait for prepared XA transaction to complete. */
+  if (wsrep_thd_is_XA_prepared(granted_thd))
+      return 1;
+
+  /* Granted is not BF, may BF abort it. */
+  if (!wsrep_thd_is_BF(granted_thd, false))
+      return 0;
+
+  /* Applying SR cannot BF abort other high priority (BF). */
+  if (wsrep_thd_is_SR(request_thd))
+      return 1;
+
+  /* Requester is truly BF and granted is applying SR in progress. */
+  if (wsrep_thd_is_SR(granted_thd))
+      return 0;
+
+  /* BF-BF wait. */
```

Comment:
Maybe you could move the error reporting here, to avoid -1 / 0 / 1 return codes?

Reply:
Yes, it's possible if we move all the BF detection logic to `lock_has_to_wait()`, specifically, to `lock_rec_has_to_wait_wsrep()`. See my comment below.